lib/index.js

```diff
@@ -1,41 +1,156 @@
+const { Stream } = require('stream')
+
 // Packages
-const getPort = require('get-port')
+const getRawBody = require('raw-body')
+const typer = require('media-typer')
+
+const { NODE_ENV } = process.env
+const DEV = NODE_ENV === 'development'
+
+const serve = fn => server((req, res) => exports.run(req, res, fn))
 
-// Ours
-const serve = require('./server')
-const listening = require('./listening')
-const getModule = require('./module')
+module.exports = serve
+exports = serve
+exports.default = serve
 
-module.exports = async (file, flags, module = getModule(file)) => {
-  const server = serve(module)
+const createError = (code, message, original) => {
+  const err = new Error(message)
 
-  let port = flags.port
-  let host = flags.host
+  err.statusCode = code
+  err.originalError = original
+
+  return err
+}
 
-  const open = await getPort(port)
-  let inUse = open !== port
+const send = (res, code, obj = null) => {
+  res.statusCode = code
 
-  if (inUse) {
-    port = open
+  if (obj === null) {
+    res.end()
+    return
+  }
 
-    inUse = {
-      old: flags.port,
-      open
+  if (Buffer.isBuffer(obj)) {
+    if (!res.getHeader('Content-Type')) {
+      res.setHeader('Content-Type', 'application/octet-stream')
     }
+
+    res.setHeader('Content-Length', obj.length)
+    res.end(obj)
+    return
   }
 
-  if (host === '0.0.0.0') {
-    host = null
+  if (obj instanceof Stream) {
+    if (!res.getHeader('Content-Type')) {
+      res.setHeader('Content-Type', 'application/octet-stream')
+    }
+
+    obj.pipe(res)
+    return
   }
 
-  server.on('error', err => {
-    console.error('micro:', err.stack)
+  let str = obj
 
-    // eslint-disable-next-line unicorn/no-process-exit
-    process.exit(1)
-  })
+  if (typeof obj === 'object' || typeof obj === 'number') {
+    // We stringify before setting the header
+    // in case `JSON.stringify` throws and a
+    // 500 has to be sent instead
 
-  server.listen(port, host, () => {
-    return listening(server, inUse, flags.silent)
-  })
+    // the `JSON.stringify` call is split into
+    // two cases as `JSON.stringify` is optimized
+    // in V8 if called with only one argument
+    if (DEV) {
+      str = JSON.stringify(obj, null, 2)
+    } else {
+      str = JSON.stringify(obj)
+    }
+
+    if (!res.getHeader('Content-Type')) {
+      res.setHeader('Content-Type', 'application/json')
+    }
+  }
+
+  res.setHeader('Content-Length', Buffer.byteLength(str))
+  res.end(str)
+}
+
+const sendError = (req, res, { statusCode, status, message, stack }) => {
+  statusCode = statusCode || status
+
+  if (statusCode) {
+    send(res, statusCode, DEV ? stack : message)
+  } else {
+    send(res, 500, DEV ? stack : 'Internal Server Error')
+  }
+
+  console.error(stack)
+}
+
+exports.send = send
+exports.sendError = sendError
+exports.createError = createError
+
+exports.run = (req, res, fn) =>
+  new Promise(resolve => resolve(fn(req, res)))
+    .then(val => {
+      if (val === null) {
```

Comment:
@matheuss depending on how much of my ideas Leo has used it should be super easy to only support it in micro-dev 🙌